libevmasm/Assembly.cpp

```diff
@@ -817,7 +830,9 @@ std::map<u256, u256> const& Assembly::optimiseInternal(
 				else
 					copy(orig, iter, back_inserter(optimisedItems));
 			}
-			if (optimisedItems.size() < m_items.size())
+			auto optimisedItemsCost = static_cast<bigint>(optimisedItems.size()) + runs * runGas(optimisedItems, KnownState{}).value;
+			auto itemsCost = static_cast<bigint>(m_items.size()) + runs * runGas(m_items, KnownState{}).value;
+			if (optimisedItemsCost < itemsCost)
```

Comment:
I think that we should keep the old condition here. It seems to me that the purpose of the check was not to ensure that new items are not more expensive - that will always be the case - but just to avoid bumping `count` if we did not replace anything. The old condition still accomplishes that and is much simpler (and cheaper to calculate).

In fact, the new condition could actually be wrong in some cases. `KnownState` you're passing in is different (empty here), so theoretically you could even get inconsistent results.

Reply:
Actually, the way `count` is being bumped here seems wrong in the first place. The loop is bumping it for every replacement, which makes sense only if the condition will always pass when there are replacements. And bumping it again in the condition makes no sense - it's not an additional optimization, just an application of already performed optimizations. It's not broken in practice only because the outer loop only cares whether it's zero or not. I think we'd be better off removing the bump and the condition and do the assignment unconditionally.